transport/tchannel/outbound.go

```diff
@@ -176,9 +179,24 @@ func callWithPeer(ctx context.Context, req *transport.Request, peer *tchannel.Pe
 		return nil, err
 	}
 
-	buf := bytes.NewBuffer(make([]byte, 0, _defaultBufferSize))
-	if _, err = buf.ReadFrom(resBody); err != nil {
-		return nil, err
+	var body io.ReadCloser
+	var bodySize int
+	if USE_BUFFER_POOL {
+		// new implementation
+		buffer := bufferpool.NewAutoReleaseBuffer()
+		if _, err = buffer.ReadFrom(resBody); err != nil {
+			return nil, err
+		}
+		body = delegateReadCloserReaderAt{buffer, bytes.NewReader(buffer.Bytes())}
```

Comment:
Few things I want to highlight: 
* we intentionally use a no-op closer at this point as wired implementation refers to the response body reader beyond the life of the `Call` handler in thrift encoding (wire.Envelope holds reference to the reader)
* Envelope unwraps the [lists lazily](https://github.com/thriftrw/thriftrw-go/blob/dev/protocol/binary/lazy_list.go#L49) where  original reader can be read late from the application code by moving around through seeker.

With this change we return a reader which on close will be put to reuse, which would will corrupt the data of the envelopes holding the reader. So this change is 100% no-go for wire clients. But such clients barely exist, so first thing we need to do would be clean up the wire code and release that version on monorepo.

Also, as no-wire has hidden expectations from reader for improved performance (which are unfortunately not clear as we didn't do good job at writing these down somewhere). No-wire needs seekers and few more things iirc, for efficiently moving around data. 

Overall this is a very risky change and can impact correctness sometimes. We should revisit this change after cleaning up thrift wired implementation and roll this buffer reuse change behind a flag

Reply:
Rephrasing to confirm I understand correctly. There are 2 thrift encoding implementations - wire (legacy) and no-wire.
- wire (legacy). This implementation holds a reference to the response body reader and its underlying []byte even _after_ closing it. lazy list is one example where the reference is maintained. Due to this existing anti-pattern, reusing the underlying []byte would be risky. 
- no-wire. There are no known interoperability issues with this implementation.

Is it possible to gate this optimization to no-wire clients only? Any other suggestion how we can realize these performance benefits in H2 2023?

Reply:
@jronak shared the following suggestion on slack:
> We can roll this change out behind an outbound option, but we need to be explicit that it is a risky change and beta change. Initially you could enable them on your services and if all looks good we can enable this by default after wire cleanup.

I've updated the PR to accommodate his suggestion.